datafusion/core/src/physical_plan/topk/mod.rs

```diff
@@ -0,0 +1,516 @@
+    /// Reservation
+    reservation: MemoryReservation,
+    /// The target number of rows for output batches
+    batch_size: usize,
+    /// sort expressions
+    expr: Arc<[PhysicalSortExpr]>,
+    /// row converter, for sort keys
+    row_converter: RowConverter,
+    /// scratch space for converting rows
+    scratch_rows: Rows,
+    /// stores the top k values and their sort key values, in order
+    heap: TopKHeap,
+}
+
+impl TopK {
+    /// Create a new [`TopK`] that stores the top `k` values, as
+    /// defined by the sort expressions in `expr`.
+    // TOOD: make a builder or some other nicer API to avoid the
+    // clippy warning
+    #[allow(clippy::too_many_arguments)]
+    pub fn try_new(
+        partition_id: usize,
+        schema: SchemaRef,
+        expr: Vec<PhysicalSortExpr>,
+        k: usize,
+        batch_size: usize,
+        runtime: Arc<RuntimeEnv>,
+        metrics: &ExecutionPlanMetricsSet,
+        partition: usize,
+    ) -> Result<Self> {
+        let reservation = MemoryConsumer::new(format!("TopK[{partition_id}]"))
+            .register(&runtime.memory_pool);
+
+        let expr: Arc<[PhysicalSortExpr]> = expr.into();
+
+        let sort_fields: Vec<_> = expr
+            .iter()
+            .map(|e| {
+                Ok(SortField::new_with_options(
+                    e.expr.data_type(&schema)?,
+                    e.options,
+                ))
+            })
+            .collect::<Result<_>>()?;
+
+        let row_converter = RowConverter::new(sort_fields)?;
+        let scratch_rows = row_converter.empty_rows(
+            batch_size,
+            20 * batch_size, // guestimate 20 bytes per row
+        );
+
+        Ok(Self {
+            schema,
+            metrics: TopKMetrics::new(metrics, partition),
+            reservation,
+            batch_size,
+            expr,
+            row_converter,
+            scratch_rows,
+            heap: TopKHeap::new(k),
+        })
+    }
+
+    /// Insert `batch`, remembering it if any of its values are among
+    /// the top k seen so far.
+    pub fn insert_batch(&mut self, batch: RecordBatch) -> Result<()> {
+        // Updates on drop
+        let _timer = self.metrics.baseline.elapsed_compute().timer();
+
+        let sort_keys: Vec<ArrayRef> = self
+            .expr
+            .iter()
+            .map(|expr| {
+                let value = expr.expr.evaluate(&batch)?;
+                Ok(value.into_array(batch.num_rows()))
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        // reuse existing `Rows` to avoid reallocations
+        let rows = &mut self.scratch_rows;
+        rows.clear();
+        self.row_converter.append(rows, &sort_keys)?;
+
+        // TODO make this algorithmically better?:
+        // 1. only check topk values in rows
+        // 2. only do one update through top_k
+
+        let mut batch_entry = self.heap.register_batch(batch);
+        for (index, row) in rows.iter().enumerate() {
```

Comment:
The is the core of the algorithm -- it is pretty unfancy but avoids allocations unless there is an actual new top value. I will file some potential arrow-rs improvements to make this even faster

Reply:
What sort of data are you testing with? I did some modeling in python so I could have a "realistic" fake data generator, and using "fail fast" strategies like this seem to provide a ~10x performance boost.

However, when query planning this, should we take into account worst-case scenarios? Because that is where I've spent my time struggling with my PR. 

When confronted with worst-case data, I can't seem to do better than being 40% worse than the existing aggregate->sort->limit.

Reply:
This is the distribution I've seen in our data:

![image](https://github.com/apache/arrow-datafusion/assets/3855243/681f4ec7-bc5e-4daa-8b6d-554aa8d7aea3)

The mean is ~0.1. So for most records time isn't advancing at all (in milliseconds), it's frequently regressing too, but the curve is shifted slightly to the right so overall time marches on. I assume this is a typical case in time-series data bases.

But if these operators get invoked on other sorts of data, or if time data was clean and continually advancing, it would could be worst case and possibly cause performance regressions if we haven't been accounting for that.

Reply:
> However, when query planning this, should we take into account worst-case scenarios? Because that is where I've spent my time struggling with my PR.

Yes I think we should  -- worst case is going to be reverse sorted data (where every value is a new TopK I think

For this particular PR I think with some more work this approach can do better in all cases than existing Sort w/ `fetch` partly due to how Sort w/ `fetch` is implemented. I do think there will need to be some decision of when to fall back to sort with limit as the sort with limit handles spilling to disk.

Reply:
FWIW, I now have an implementation of the limited aggregator that seems to be always faster than sort with limit, at least for small limits. I think I can also make it ~40% faster with research I have in other branches.

So in short, I no longer think we need to fallback, at least if we spend enough time optimizing. Also of note for testing: my data generator does simulated timeseries data, or worst case with a flag, if you find this helpful for your work.

Reply:
It is [here](https://github.com/avantgardnerio/arrow-datafusion/blob/5d3268b0cf8cdeeab0049f3d96c4e120afa10324/datafusion/core/src/physical_plan/aggregates/priority_map.rs#L564). I can PR it separately if you think it's useful.

It does things like look at real distributions for number of records per trace:

![image](https://github.com/apache/arrow-datafusion/assets/3855243/8bb66f7d-73b4-44a4-88ea-2b993aa1fbb1)

and approximate it with a Pareto distribution:

![image](https://github.com/apache/arrow-datafusion/assets/3855243/f9a0f1bf-4d6c-446e-b1de-f4469ae02cde)

and similar for timings, so I think it exercises branches approximately the real way time series data would.